[blog] Add Demucs ONNX export blog

dhunstack opened the pull request:
Blog documenting the "Converting Demucs v4 (Hybrid Transformer) AI model to ONNX format" GSOC project.

[preview](https://mixxx.org/news/2025-10-27-gsoc2025-demucs-to-onnx-dhunstack/)

JoergAtGithub commented:
Please fix the pre-commit issues first (additional space at line end and missing new line at file end)

JoergAtGithub commented on content/news/XXXX-XX-XX-gsoc2025-demucs-to-onnx-dhunstack.md:
```diff
@@ -0,0 +1,99 @@
+title: GSOC 2025 - Converting Demucs v4 (Hybrid Transformer) AI model to ONNX format
+authors: Anmol Mishra
+tags: gsoc, gsoc-2025, stems
```
```suggestion
authors: Anmol Mishra
status: draft
tags: gsoc, gsoc-2025, stems
```
I think the preview does not build, because this statement is missing

JoergAtGithub commented on content/news/XXXX-XX-XX-gsoc2025-demucs-to-onnx-dhunstack.md:
```diff
@@ -0,0 +1,100 @@
+
+Disclaimer: *This blog post primarily serves as the documentation for the [Google Summer of Code](https://summerofcode.withgoogle.com/programs/2025/projects/lRQpeA7K) 2025 project: "Converting Demucs v4 (Hybrid Transformer) AI model to ONNX format".*
+
+Mixxx currently supports playback of stem files, but does not support building stems from scratch. This project was a step towards supporting real time stem inference from within Mixxx. The key contributions of this project are -
```
```suggestion
Mixxx 2.6 will support playback of stem files, but does not support building stems from scratch. This project was a step towards supporting real time stem inference from within Mixxx. The key contributions of this project are -
```
Stem files can be created from the original seperated tracks recorded in a studio (or from DAW) or build by stem seperation out of any track. You should be clear what this is about.
And add a link to https://www.stems-music.com/ that the reader can find out what the Stem file is.

JoergAtGithub replied:
The reader need more introduction to understand the following text. Explain especially what Demucs is.

JoergAtGithub commented on content/news/XXXX-XX-XX-gsoc2025-demucs-to-onnx-dhunstack.md:
```diff
@@ -0,0 +1,100 @@
+
+Source separation models are typically developed using Python-based libraries such as PyTorch or TensorFlow, which require a Python runtime to execute exported machine learning models. This dependency can make it challenging to use these models in environments where Python is not available or practical, such as audio plugins, which are often written in C++. To address this, ONNX (Open Neural Netw ...
+
+While both PyTorch and TensorFlow provide straightforward ONNX export for simple models, complex audio models like Demucs often present significant hurdles. This is primarily due to their use of complex tensors and custom implementations of operations like STFT and ISTFT, which are not natively supported by ONNX. As a result, previous attempts to export Demucs to ONNX have typically bypassed thes ...
```
Explain what STFT and ISTFT are.
Also explain why you could reduce the fourier transformation from complex to the real part only.

JoergAtGithub commented on content/news/XXXX-XX-XX-gsoc2025-demucs-to-onnx-dhunstack.md:
```diff
@@ -0,0 +1,100 @@
+- [feat] Add ONNX exportable STFT implementation — [mixxxdj/demucs#3](https://github.com/mixxxdj/demucs/pull/3)
+- [feat] Add ONNX exportable ISTFT implementation — [mixxxdj/demucs#4](https://github.com/mixxxdj/demucs/pull/4)
+- Add ONNX export instructions — [mixxxdj/demucs#5](https://github.com/mixxxdj/demucs/pull/5)
+
```
You should mention upstreaming the changes to https://github.com/adefossez/demucs once you opened the PR

Also mention your upcoming talk at ADC2025

daschuer commented on content/news/XXXX-XX-XX-gsoc2025-demucs-to-onnx-dhunstack.md:
```diff
@@ -0,0 +1,283 @@
+
+Disclaimer: *This blog post primarily serves as the documentation for the [Google Summer of Code](https://summerofcode.withgoogle.com/programs/2025/projects/lRQpeA7K) 2025 project: "Converting Demucs v4 (Hybrid Transformer) AI model to ONNX format".*
+
+Mixxx 2.6 will support playback of [stem files]( https://www.stems-music.com/), which can be created from the original track stems recorded in a studio (or DAW), or by using AI based stem separation models on the final mixed track. [Demucs v4](https://github.com/adefossez/demucs) is a state-of-the-art open-source music source separation model developed by Antoine Défossez. While it provides excep ...
```
```suggestion
Mixxx 2.6 will support playback of [stem files](https://www.stems-music.com/), which can be created from the original track stems recorded in a studio (or DAW), or by using AI based stem separation models on the final mixed track. [Demucs v4](https://github.com/adefossez/demucs) is a state-of-the-art open-source music source separation model developed by Antoine Défossez. While it provides exceptional quality in separating audio into stems, it is currently implemented in Python and therefore cannot be used directly in C++ applications or run efficiently on hardware accelerators through [ONNX Runtime](https://onnxruntime.ai/). 
```

daschuer replied:
A brief word explaining stem files would be nice to not loose readers in the first paragraph:  

support playback of [stem files](https://www.stems-music.com/) with contain separate channels for different instruments.

daschuer replied:
Let's use the Mixxx user perspective. Something like: 

.. which can obtained from from the original distribution, or if not available by one of the AI based splitter tools like ...

daschuer replied:
Demucs v4 can make use of its own paragraph. With an additional sentences how it works for non tech readers. 

Is "ONNX Runtime" the only option or shall we mention alternatives and than give a reason why ONNX Runtime works best for us.

daschuer replied:
There is a mix up with  "ONNX Runtime" and the ONNX format. Maybe we can mention runtimes that support that format.

daschuer commented on content/news/XXXX-XX-XX-gsoc2025-demucs-to-onnx-dhunstack.md:
```diff
@@ -0,0 +1,283 @@
+
+Mixxx 2.6 will support playback of [stem files]( https://www.stems-music.com/), which can be created from the original track stems recorded in a studio (or DAW), or by using AI based stem separation models on the final mixed track. [Demucs v4](https://github.com/adefossez/demucs) is a state-of-the-art open-source music source separation model developed by Antoine Défossez. While it provides excep ...
+
+This project is a step towards supporting real time stem separation within Mixxx, by exporting the Python based Demucs model to [ONNX](https://onnx.ai/) (Open Neural Network Exchange format). The key contributions of this project are -
```
```suggestion
This project is a step towards supporting real time stem separation within Mixxx, by exporting the Python based Demucs model to [ONNX](https://onnx.ai/) (Open Neural Network Exchange format). The key contributions of this project are:
```

daschuer commented on content/news/XXXX-XX-XX-gsoc2025-demucs-to-onnx-dhunstack.md:
```diff
@@ -0,0 +1,283 @@
+
+- Preparing the Demucs code for ONNX export by rewriting non-exportable operations
+- Validating all the modified parts with numerical tests to ensure export of existing model weights without the need for retraining.
+- Scripts for exporting ONNX and ORT formats of the model.
```
What is ORT? Can you mention it in the paragraph above?

JoergAtGithub replied:
ORT == ONNX Runtime: https://onnxruntime.ai/
Not self explaining for readers without background knowledge.

daschuer commented on content/news/XXXX-XX-XX-gsoc2025-demucs-to-onnx-dhunstack.md:
```diff
@@ -0,0 +1,283 @@
+
+While both PyTorch and TensorFlow provide straightforward ONNX export for simple models, complex audio models like Demucs often present significant hurdles. This is primarily due to their use of complex tensors and custom implementations of operations like STFT and ISTFT, which are not natively supported by ONNX. As a result, previous attempts to export Demucs to ONNX have typically bypassed thes ...
+
+In this project, we successfully exported Demucs v4 as a fully self-contained ONNX model, including the STFT and ISTFT layers. This means the model can be used directly in any language that supports ONNX, greatly simplifying integration for audio developers and enabling broader adoption of source separation technology.
```
... including the STFT and ISTFT layers

Can you explain this a bit more. Does ONNX now support that natively?

JoergAtGithub replied:
I think this sentence is clear and correct. The details of the implementation are described in the paragraphs below.

daschuer replied:
My issue is that I have no clue about layers in an ONNX model. This is not described anywhere else in this blog post. So you may describe it differently or just remove this detail..

dhunstack replied:
A `Layer` in the context of Deep Learning can be thought of as an abstracted operation.

DL evolved over time to have more complicated operations, so researchers started combining a standard set of fundamental operations together and created more abstract "layers", mainly for easier communication and concise description.
STFT and ISTFT can also be considered layers as they're a set of operations that transform the input in a precisely defined way. 

Large Deep Learning models are thus nothing but a set of these chained layers, however individual layers have become complex over time as they now combine other more simple layers. These more complex layers are also known as `blocks`.

In the text, I've used convolution and Transformer layers/blocks in the Demucs explanation.

But I'll remove this term here, will use "STFT and ISTFT operations" which might be better understood.

daschuer commented on content/news/XXXX-XX-XX-gsoc2025-demucs-to-onnx-dhunstack.md:
```diff
@@ -0,0 +1,283 @@
+
+In this project, we successfully exported Demucs v4 as a fully self-contained ONNX model, including the STFT and ISTFT layers. This means the model can be used directly in any language that supports ONNX, greatly simplifying integration for audio developers and enabling broader adoption of source separation technology.
+
+A key part of our approach was to diagnose and address each export issue step by step, developing custom numerical tests to ensure that our ONNX-compatible replacements matched the original PyTorch calculations as closely as possible. This was crucial, as retraining the model was not a viable option: Demucs was originally trained on a large private dataset over weeks using multiple GPUs, and repr ...
```
Why only "closely" and not "bit perfect"?

dhunstack replied:
Because PyTorch's STFT and ISTFT are implemented in C++ (and assembly). 
For ONNX export, I could only use the subset of PyTorch operations that are ONNX exportable, so the operations used can differ. Depending on how those operations are ultimately implemented, numerical differences could have occurred.
Which is why we did stringent tests, and benchmarks to evaluate the exported model.

daschuer commented on content/news/XXXX-XX-XX-gsoc2025-demucs-to-onnx-dhunstack.md:
```diff
@@ -0,0 +1,283 @@
+Example: a 44.1 kHz stereo track stores **44,100 real values per second per channel**.  
+Each value (e.g. `0.25`, `-0.67`) represents instantaneous amplitude (air pressure or voltage).
+
+#### Complex Numbers (Frequency Representation)
```
This is to fast. Something like this is missing: 

A complex number directed vector that can be the amplitude and a phase of a frequency in a track.

daschuer replied:
I wonder if that details contributes to the blog post. Maybe you can explain it implicit along with the FFT part below.

JoergAtGithub replied:
I think this helps to understand in the paragraphs below to understand why only the real part was necessary.

daschuer commented on content/news/XXXX-XX-XX-gsoc2025-demucs-to-onnx-dhunstack.md:
```diff
@@ -0,0 +1,283 @@
+- Operates directly on the raw waveform (real-valued input).  
+- Uses temporal convolutions and Transformer blocks to learn how each instrument evolves over time.
+
+##### **Time-Frequency Domain Branch**  
```
I am missing a paragraph like this (my rookie understanding) 

The FFT is used to demangle the row of samples to the original bits of sound. Micro oscillations (sinusoid) like emitted of a certain bras sheet of a trumpet. The sum of these sinusoids can be converted back to the streams of sound samples. 

The Demux model takes all these resinoids and compares it to the training data and decide which of them belongs to which instrument. This way it can create one stream for each instrument.

dhunstack replied:
In this blog, I've tried to not talk much about the internal DL architecture, instead focusing only on the export related issues with Demucs. A discussion on the mechanisms of Demucs's layers might be too complex (and too much information) for this blog. 
I've referred to convolution and Transformer layers, whose internal mechanisms are well described mathematically but the math doesn't have an easy interpretation. Hence why the AI models are referred to as black boxes. 
I can link to the actual research paper though, if you suggest that'd be helpful - https://arxiv.org/pdf/2211.08553

daschuer replied:
Your new version reads already good. Thank you.

daschuer commented on content/news/XXXX-XX-XX-gsoc2025-demucs-to-onnx-dhunstack.md:
```diff
@@ -0,0 +1,283 @@
+
+#### Real-Valued STFT and ISTFT Rewrites
+
+Because complex values appear **only** in the initial STFT and final ISTFT layers, our solution was to **reimplement these operations** in a real-valued form.
```
This does not sound correct. Isn't the sine/coin notation just a different notation of the same complex value?

JoergAtGithub replied:
I think it is correct and uses the common mathematical terms. You could use the polar representation (magnitude and phase) instead of real-value + imaginary value. But to explain this work the later fits better.
sine/cosine are operations - not a representation.
Maybe a simple drawing like https://en.wikipedia.org/wiki/Complex_number#/media/File:Complex_number_illustration_modarg.svg would help the reader to understand the difference between magnitude, amplitude and phase?

daschuer replied:
I don't struggling to understand the complex number itself. 

For my understanding the sentences is misleading, because the complex-valued nature of the operation must still exists conceptually. It looks like the rewritten STFT avoids using complex tensors, by using a sinus and a cosinus temsor pair that is representing the same info. 

z=a+bi=reiθ=r(cosθ+isinθ)

Is that understanding correct?

dhunstack replied:
I see. Yeah the underlying math is same, but the implementation performs the operations separately for real part and imaginary part, so avoiding any complex tensor data type.
Would you think this statement might be a better fit?
`
Because complex tensors appear **only** in the initial STFT and final ISTFT layers, our solution was to **reimplement these operations** using only real tensors.
`

daschuer replied:
Your explanation is very good, thank you. Unfortunately it is not integrated in the proposal. Here a try: 

```suggestion
Because complex tensors appear **only** in the initial STFT and final ISTFT layers, our solution was to **reimplement these operations** using separate real tensors for the real and the imagine part of the original complex values, while keeping the underlying math unchanged.
```

daschuer commented on content/news/XXXX-XX-XX-gsoc2025-demucs-to-onnx-dhunstack.md:
```diff
@@ -0,0 +1,283 @@
+- The original ISTFT combines complex values through real + imaginary synthesis.
+- We reconstructed the time-domain signal by performing the same series of **overlap-add** and **cosine/sine inverse convolutions**, again using only real-valued tensors.
+
+By carefully ensuring numerical equivalence to PyTorch’s implementation, we achieved perfect parity (MSE < 1e-4) between the original and rewritten layers.
```
nice :-)

daschuer commented on content/news/XXXX-XX-XX-gsoc2025-demucs-to-onnx-dhunstack.md:
```diff
@@ -0,0 +1,283 @@
+
+## Benchmark Results
+
+The quality of the model is expected to be equal or slightly worse when exported to ONNX. While there are plenty of ways of measuring the benchmarks models (another blog post incoming), we've chosen to measure our models with `SI-SDR` metric, Scale Invariant Signal To Distortion Ratio, on the MusDB dataset. This is the standard metric on which researchers report their source separation model's pe ...
```
Can you add other sentence with example db values? What would be perfect what is expected?

JoergAtGithub replied:
The numerical values for Demucs are in the table below. I don't think additional numbers will improve the text here.

daschuer replied:
I struggle to understand which value is better. Is 0 dB the optimum?

dhunstack replied:
@daschuer 
I can probably add a line here that higher dB values are better. 
This page is a great way to get familiarized with the SI-SDR metric - https://source-separation.github.io/tutorial/basics/evaluation.html
This is a table for SDR evaluation of Demucs compared to other models - https://github.com/facebookresearch/demucs?tab=readme-ov-file#comparison-of-accuracy

The evaluation of source separated tracks is an important topic in itself. The above URL I've mentioned talks about evaluation of ML models, but doesn't discuss the topic of quality assessment for exported models, which has multiple potential sources of quality degradation. Which is why I've added the comment for `(another blog post incoming)` in the text.

daschuer replied:
That works for me. Thank you.

Interesting links. You may add another sentence summarisimg your good explanations here.

daschuer replied:
Idea: for the subject part you may add two files for compassion python vs. c++.

daschuer commented on content/news/XXXX-XX-XX-gsoc2025-demucs-to-onnx-dhunstack.md:
```diff
@@ -0,0 +1,283 @@
+math: yes
+comments: yes
+
+Disclaimer: *This blog post primarily serves as the documentation for the [Google Summer of Code](https://summerofcode.withgoogle.com/programs/2025/projects/lRQpeA7K) 2025 project: "Converting Demucs v4 (Hybrid Transformer) AI model to ONNX format".*
```
It would be great to have an eye catching vision statement at the very beginning. Something like: 

Imagine you could load a track and .....

daschuer reviewed (changes requested):
Wow, this is really impressive. Thank you. 
I have left some comments to make it more accessible for people not that deep into the AI topic.

daschuer commented:
This type of AI has nothing to do with Chat-GPT. I can imagine that people do not know this. So maybe you may ad short paragraph about pattern recognition. And a brief compare about language models and Audio models.

JoergAtGithub commented on content/news/XXXX-XX-XX-gsoc2025-demucs-to-onnx-dhunstack.md:
```diff
@@ -0,0 +1,283 @@
+
+Disclaimer: *This blog post primarily serves as the documentation for the [Google Summer of Code](https://summerofcode.withgoogle.com/programs/2025/projects/lRQpeA7K) 2025 project: "Converting Demucs v4 (Hybrid Transformer) AI model to ONNX format".*
+
+Mixxx 2.6 will support playback of [stem files]( https://www.stems-music.com/), which can be created from the original track stems recorded in a studio (or DAW), or by using AI based stem separation models on the final mixed track. [Demucs v4](https://github.com/adefossez/demucs) is a state-of-the-art open-source music source separation model developed by Antoine Défossez. While it provides excep ...
```
```suggestion
Mixxx 2.6 will support playback of [stem files]( https://www.stems-music.com/), which can be created from the original track stems recorded in a studio (or DAW), or by using AI based stem separation models on the final mixed track. [Demucs v4](https://github.com/adefossez/demucs) is a state-of-the-art open-source music source separation model developed by Alexandre Défossez. While it provides exceptional quality in separating audio into stems, it is currently implemented in Python and therefore cannot be used directly in C++ applications or run efficiently on hardware accelerators through [ONNX Runtime](https://onnxruntime.ai/). 
```

JoergAtGithub commented on content/news/XXXX-XX-XX-gsoc2025-demucs-to-onnx-dhunstack.md:
```diff
@@ -0,0 +1,283 @@
+
+To understand the export process, it’s important to look closely at how **HTDemucs (Hybrid Transformer Demucs)** is structured, and why its architecture introduces challenges when converting to the **ONNX** format.
+
+![htdemucs]({static}/images/news/demucs-to-onnx/htdemucs.png)
```
For copyright reasons please add the information where you've taken the image from and as the file is taken from the Demucs project under MIT license, you also need to link the license file.

Eve00000 commented:
impressive! thank you.

ywwg commented:
This project is amazing. This is real, groundbreaking work being done in Mixxx!  I love that most of it is merged already, too :)

daschuer commented on content/news/XXXX-XX-XX-gsoc2025-demucs-to-onnx-dhunstack.md:
```diff
@@ -0,0 +1,283 @@
+Where:
+
+- $r = \sqrt{a^2 + b^2}$ (magnitude)  
+- $\theta = \operatorname{atan2}(b, a)$ (phase angle)
```
I get this warning on GitHub: 
> The following macros are not allowed: operatorname

Not sure if this is a real issue.

dhunstack commented:
@daschuer @JoergAtGithub I've edited in your suggestions in the blog

Eve00000 commented:
@dhunstack, I hope you understand that it will take me some time to really understand what & how you did it.
I,m very curious and keen to test out your work but I’m already convinced our users will love what you created.

acolombier commented on content/news/XXXX-XX-XX-gsoc2025-demucs-to-onnx-dhunstack.md:
```diff
@@ -0,0 +1,312 @@
+
+Source separation models are typically developed using Python-based libraries such as PyTorch or TensorFlow, which require a Python runtime to execute machine learning models. This dependency can make it challenging to use these models in environments where Python is not available or practical, such as audio plugins, which are often written in C++. To address this, ONNX (Open Neural Network Excha ...
+
+While both PyTorch and TensorFlow provide straightforward ONNX export for simple models, complex audio models like Demucs often present significant hurdles. This is primarily due to their use of complex tensors and custom implementations of operations like STFT and ISTFT, which are not natively supported by ONNX. As a result, previous attempts to export Demucs to ONNX have typically bypassed thes ...
```
Similarly to how you've done it when introducing the `ONNX` acronym, it would be great to express what STFT and ISTFT stand for, and also potentially link introduction that what they are (eg. [Wikipedia for STFT](https://en.wikipedia.org/wiki/Short-time_Fourier_transform))

Edit: you do it later, but IMO it is great to give the full meaning the first time you use the acronym

acolombier reviewed (approved):
This is looking already very good! Just a small nit, but otherwise I think this is ready to go!

daschuer reviewed (approved):
I am really exited about your project after reading this Blog. Thank you. LGTM.

daschuer commented:
@acolombier Can you take care for merging? @dhunstack needs a stable link for his submission before the deadline, tomorrow October 27, 2025 8 PM CET.

JoergAtGithub commented on content/news/XXXX-XX-XX-gsoc2025-demucs-to-onnx-dhunstack.md:
```diff
@@ -0,0 +1,312 @@
+| Example C++ Scripts   |      In Review      |     [PR #7](https://github.com/mixxxdj/demucs/pull/7)    |
+| Benchmarking Scripts  |      In Review      |     [PR #9](https://github.com/mixxxdj/demucs/pull/8)    |
+
+We've raised a PR with all our Demucs changes to the upstream here - [ADD LINK].
```
Here is still a placeholder for the link!

dhunstack committed:
[blog] Add Demucs ONNX export blog

Blog documenting the "Converting Demucs v4 (Hybrid Transformer) AI model
to ONNX format" GSOC project.

Signed-off-by: Anmol Mishra <anmolmishra1997@gmail.com>

JoergAtGithub commented on content/news/XXXX-XX-XX-gsoc2025-demucs-to-onnx-dhunstack.md:
```diff
@@ -0,0 +1,313 @@
+## References
+
+[1] A. Défossez, “Hybrid Transformers for Music Source Separation,” *arXiv preprint arXiv:2211.08553*, 2022.  
+Available: https://arxiv.org/abs/2211.08553
```
```suggestion
Available: [https://arxiv.org/abs/2211.08553](https://arxiv.org/abs/2211.08553)
```

JoergAtGithub commented on content/news/XXXX-XX-XX-gsoc2025-demucs-to-onnx-dhunstack.md:
```diff
@@ -0,0 +1,313 @@
+
+Source separation models are typically developed using Python-based libraries such as PyTorch or TensorFlow, which require a Python runtime to execute machine learning models. This dependency can make it challenging to use these models in environments where Python is not available or practical, such as audio plugins, which are often written in C++. To address this, ONNX (Open Neural Network Excha ...
+
+While both PyTorch and TensorFlow provide straightforward ONNX export for simple models, complex audio models like Demucs often present significant hurdles. This is primarily due to their use of complex tensors and custom implementations of operations like Short-time Fourier Transform ([STFT](https://en.wikipedia.org/wiki/Short-time_Fourier_transform)) and Inverse Short-time Fourier Transform ([I ...
```
```suggestion
While both PyTorch and TensorFlow provide straightforward ONNX export for simple models, complex audio models like Demucs often present significant hurdles. This is primarily due to their use of complex tensors and custom implementations of operations like Short-time Fourier Transform ([STFT](https://en.wikipedia.org/wiki/Short-time_Fourier_transform)) and Inverse Short-time Fourier Transform ([ISTFT](https://en.wikipedia.org/wiki/Short-time_Fourier_transform#Inverse_STFT)), which are not natively supported by ONNX. As a result, previous attempts to export Demucs to ONNX have typically bypassed these operations, requiring developers to reimplement STFT/ISTFT in every target language, a process that is both error prone and time consuming.
```

JoergAtGithub commented on content/news/XXXX-XX-XX-gsoc2025-demucs-to-onnx-dhunstack.md:
```diff
@@ -0,0 +1,313 @@
+| ONNX Export Scripts   |       Merged        |     [PR #5](https://github.com/mixxxdj/demucs/pull/5)    |
+| CI for model export   |       Merged        |     [PR #6](https://github.com/mixxxdj/demucs/pull/6)    |
+| Example C++ Scripts   |      In Review      |     [PR #7](https://github.com/mixxxdj/demucs/pull/7)    |
+| Benchmarking Scripts  |      In Review      |     [PR #9](https://github.com/mixxxdj/demucs/pull/8)    |
```
The benchmark PR #9 is missing and the link points to #8

JoergAtGithub commented:
Thank you very much for this great contribution! It was a pleasure to work with you on this GSoC project!
